providers/implementations/digests/ml_dsa_mu_prov.c

```diff
@@ -0,0 +1,339 @@
+ * https://www.openssl.org/source/license.html
+ */
+
+/*
+ * Mu is the value:
+ *  mu = SHAKE256(tr || M', 64)
+ *
+ * where tr is the hash of the public key
+ * And M' is one of the following:
+ *   (1) Pure: M' = 00 || ctx_len || ctx || in (where in = message)
+ *   (2) PreHash: M' = 01 || ctx_len || ctx || OID || in (where in = hashed(msg))
+ *
+ * M` = 01 || ctx_len || ctx || OID || PH(M)
+ *
+ */
+
+#include "internal/deprecated.h" /* including crypto/sha.h requires this */
+
+/*
+ * NOTE: By default CSHAKE sets secure xof lengths (OSSL_DIGEST_PARAM_XOFLEN)
+ * that are used by EVP_DigestFinal_ex(). This differs from SHAKE where the
+ * xof length MUST be set (since the initial implementation shipped with BAD
+ * defaults - and the only safe way to fix it was to make the user set the value)
+ */
+#include <string.h>
+#include <openssl/evp.h>
+#include <openssl/err.h>
+#include <openssl/proverr.h>
+#include <openssl/core_names.h>
+#include "crypto/ml_dsa.h"
+#include "prov/provider_ctx.h"
+#include "prov/digestcommon.h"
+#include "prov/der_pq_dsa.h"
+#include "prov/implementations.h"
+#include "internal/common.h"
+#include "internal/sha3.h"
+#include "providers/implementations/digests/ml_dsa_mu_prov.inc"
+
+#define SHAKE256_SIZE 64
+#define SHAKE_FLAGS (PROV_DIGEST_FLAG_ALGID_ABSENT)
+#define ML_DSA_MAX_CONTEXT_STRING_LEN 255
+
+typedef struct mu_ctx_st {
+    OSSL_LIB_CTX *libctx;
+    char *propq;
+    EVP_MD_CTX *mdctx;
+    EVP_MD *md;
+    uint8_t context[ML_DSA_MAX_CONTEXT_STRING_LEN];
+    size_t context_len;
+    uint8_t tr[SHAKE256_SIZE];   /* Pre-cached public key Hash */
+    const uint8_t *oid;
+    size_t oid_len;
+    size_t digest_len;
+    size_t remaining;
+} MU_CTX;
+
+static OSSL_FUNC_digest_newctx_fn mu_newctx;
+static OSSL_FUNC_digest_freectx_fn mu_freectx;
+static OSSL_FUNC_digest_get_params_fn mu_get_params;
+static OSSL_FUNC_digest_dupctx_fn mu_dupctx;
+static OSSL_FUNC_digest_init_fn mu_init;
+static OSSL_FUNC_digest_update_fn mu_update;
+static OSSL_FUNC_digest_final_fn mu_final;
+static OSSL_FUNC_digest_set_ctx_params_fn mu_set_ctx_params;
+static OSSL_FUNC_digest_settable_ctx_params_fn mu_settable_ctx_params;
+static OSSL_FUNC_digest_get_ctx_params_fn mu_get_ctx_params;
+static OSSL_FUNC_digest_gettable_ctx_params_fn mu_gettable_ctx_params;
+
+static void *mu_newctx(void *provctx)
+{
+    MU_CTX *ctx;
+
+    if (ossl_unlikely(!ossl_prov_is_running()))
+        return NULL;
+    ctx = OPENSSL_zalloc(sizeof(*ctx));
+    if (ctx != NULL)
+        ctx->libctx = PROV_LIBCTX_OF(provctx);
+    return ctx;
+}
+
+static void mu_freectx(void *vctx)
+{
+    MU_CTX *ctx = (MU_CTX *)vctx;
+
+    OPENSSL_free(ctx->propq);
+    EVP_MD_free(ctx->md);
+    EVP_MD_CTX_free(ctx->mdctx);
+    OPENSSL_clear_free(ctx, sizeof(*ctx));
+}
+
+static void *mu_dupctx(void *ctx)
+{
+    MU_CTX *src = (MU_CTX *)ctx;
+    MU_CTX *dst = ossl_prov_is_running() ? OPENSSL_malloc(sizeof(*dst)) : NULL;
+    if (dst != NULL) {
```

Comment:
Perhaps instead:

```suggestion

    if (dst = NULL)
        return NULL;
```
and then de-indent the rest of the code?